### ISSUE TITLE
Support for `.xcstrings` catalog

### ISSUE THREAD

**Issue by danieleformichelli** [tuist/tuist#6048]
### What problem or need do you have?

They were added in Xcode 15, see https://github.com/tuist/tuist/discussions/5221#discussioncomment-6103709

### macOS version

14.4

### Tuist version

4.5.0

### Xcode version

15.3

**Comment by muukii** [comment]
what's the problem actually? Xcode will produce strings or stringsdict automatically if we put a xcstrings.

**Comment by leszko11** [comment]
Hey @muukii 👋🏼
Are you sure about that? I have never used strings catalogs, but after a quick test on `ios_app_with_frameworks_and_resources` I see that whenever I migrate `.strings`to a strings catalog, it doesn’t exist in a project anymore.
I think the point of this issue is to include strings catalog file in a generated workspace, because now it’s not included after generation. 
However, swiftgen does not support strings catalog yet according to [this issue](https://github.com/SwiftGen/SwiftGen/issues/1065). What should we do about it?

**Comment by muukii** [comment]
In my case, using different name to `Localizable` because I still have classic style strings file.
Specify table name as `StringCatalog` to load. Not sure it's making some effects.

![CleanShot 2024-03-14 at 13 06 23@2x](https://github.com/tuist/tuist/assets/1888355/a2a448cd-669a-4491-b135-43afcd4cb8f1)
![CleanShot 2024-03-14 at 13 07 34@2x](https://github.com/tuist/tuist/assets/1888355/00aefab1-c435-4bd6-8de3-ffa30828a3e9)

**Comment by leszko11** [comment]
Yes, that's the possible solution, but in the context of tuist capabilities, it looks a bit like a workaround to hold both of these. I think 

Just out of curiosity, are `StringCatalog.strings` and `StringCatalog.xcstrings` update each other, or if you want to add a new key, you need add it in both files?

**Comment by leszko11** [comment]
A little update on this one.

As I mentioned before _SwiftGen_ does not support `.xcstrings` files yet. 
I've tried to implement the solution mentioned [here](https://github.com/SwiftGen/SwiftGen/issues/1065#issuecomment-1773334930), however, flow `.xcstrings` -> `.strings` -> `.swift` does not fit into current resource synthesizer generator, because it requires running it twice. In addition, it doesn't support plurals(`.stringsdict`)

IMHO best solution would be to write custom `.stencil` supporting `.xcstrings` files in _SwiftGen_, and then extend `TuistGenerator` with a `.xcstrings` support. I can start to work on it, but I'm not an expert in writing stencils(yet 🤞)

**Comment by pepicrft** [comment]
> IMHO best solution would be to write custom .stencil supporting .xcstrings files in SwiftGen, and then extend TuistGenerator with a .xcstrings support. I can start to work on it, but I'm not an expert in writing stencils(yet 🤞)

@leszko11 is there any way we can support you in making this happen?

**Comment by leszko11** [comment]
@pepicrft  It's in progress. I know it's going slow, but so far I've implemented all the setup required to generate `.swift` file from `.xcstring` and started to write a stencil template. I'm trying to be as close as possible to the strings asset file generated from `.strings`.
Aiming to complete it end of this week

**Comment by fortmarek** [comment]
Amazing, thanks for the update 🙌 Let us know if you need anything from us 🙂 Happy to take a look at some early draft if you want to align on the direction of the PR

**Comment by leszko11** [comment]
I'm stuck 😔

`.xcstring` file under the hood is a `.json` file, so it needs swiftgen's [JSON parser](https://github.com/SwiftGen/SwiftGen/blob/stable/Documentation/Parsers/json.md) to process the file. Not using [strings parser](https://github.com/SwiftGen/SwiftGen/blob/stable/Documentation/SwiftGenKit%20Contexts/strings.md) means that we're losing an access to a `types` property which holds a list of arguments used in translation. Without it, I have no clue, how to do it.

I've found [this tool](https://github.com/liamnichols/xcstrings-tool), which basically uses a strings parser to process `.xcstring` file and produce synthesised resources

**Comment by pepicrft** [comment]
@leszko11 it seems that you'll have to contribute a new parser to [SwiftGen](https://github.com/SwiftGen/SwiftGen/) that might reuse logic from the parsers that you mentioned (the JSON and the Strings). I'm not entirely familiar with the SwiftGen codebase, but it seems that those parsers can be of good reference.

**Comment by leszko11** [comment]
@pepicrft Yep, I think that would be the best solution. I'll start a conversation in [corresponding issue](https://github.com/SwiftGen/SwiftGen/issues/1065)

**Comment by leszko11** [comment]
https://github.com/SwiftGen/SwiftGen/pull/1124

👀

**Comment by pepicrft** [comment]
Niiiice! SwiftGen users will be very excited about that contribution. Thanks for making it happen @leszko11

**Comment by NiklasOemler** [comment]
Commenting to stay informed about this topic. 
Can't wait to migrate to string catalogs when youre done and it's available in a new tuist version!  🎉

Greetings

**Comment by Jeehut** [comment]
Hey guys, I'm currently working in a project where Tuist is being used and I'm kinda confused. I'm very proficient with localization in iOS development and String Catalogs, but the way this discussion is going, including how the PR is trying to "add support" for String Catalogs contradicts with my whole understanding of String Catalogs.

One of the most common misconceptions about String Catalogs is that it's "just another format" compared to Strings(dict) files. But it's not. It's an entirely different paradigm. I wrote an entire [blog post](https://www.fline.dev/the-missing-string-catalogs-faq-for-xcode-15/) about String Catalogs with my learnings from chatting with the Apple engineers who built it.

The key misconception here seems to be that String Catalogs are the source of truth for keys, like Strings(dict) files were. Previously, you would write something like `NSLocalizedString("your-key-name")` and you would hope that the source of truth – which is the Strings file – would include that key. It was easy to mistype the key or the key could have been removed from the Strings file, breaking the localization. Hence, tools like SwiftGen were used to make sure only keys that actually existed in Strings files would be referenced in code, which would read something like `L10n.Your.Key.name`. (BTW, I [contributed](https://github.com/SwiftGen/SwiftGen/pull/159) this structured style to SwiftGen back then.)

But String Catalogs are different. They are NOT the source of truth. Instead, Xcode will always automatically find and extract any key appearances in code like in `NSLocalizedString("your-key-name")`, or the modern alternative `String(localized: "your-key-name")` or even any parameters of type `LocalizedStringKey` which is found everywhere in SwiftUI. And then, based on the keys it found, it will add any missing keys to the String Catalog. It will also mark any keys that it found in the String Catalog but no references to in code as "Stale". And it will show which languages have missing translations to even ensure that each key is translated in all languages and developers can easily detect missing translations.

This is a new world. And it's a great one, too. I had built a [tool](https://github.com/FlineDev/BartyCrouch) that basically did the same extraction kind of thing, but I had to work with Strings(dict) files, so my [recommended usage guide](https://jeehut.medium.com/localization-in-swift-like-a-pro-48164203afe2) was still recommending to use it in combination with SwiftGen to get full safety. But this is no longer needed. Instead, I've built a [simple drag & drop app](https://translatekit.app) for localizing String Catalogs without any safety feature provided by me even though I've always strived for improving the localization workflow in Xcode. It's all no longer needed. No need for SwiftGen, no need for any other tooling. Xcode makes our code the source of truth now – and makes sure any key exists in the String Catalog.

Now, coming back to Tuist, from my understanding all that Tuist needs to do is to provide a way to specify the locales supported by an app. Because there are no longer `en.lproj` folders needed with String Catalogs, all languages are in a single JSON structure. The String Catalog seems to get added to the target just fine in my testing, so Xcode would handle the rest. BUT: I couldn't find a way to specify in my `Project.swift` how to define which locales my app supports. I would expect there to be an array parameter named `localizations` that accepts something like `["en", "de", "fr-CA"]`. Alternatively, Tuist could automatically detect the languages in any String Catalog (`.xcstrings`) files, but then there would be questions like: What if there are 10 languages, but only 5 of them actually 100% localized, the rest still empty? What if there are multiple `.xcstrings` files with different languages?

I'm still new to Tuist, so maybe I'm just missing something and the option is already there. But I would definitely recommend and expect that I can somehow define what's listed in the following setting, cause I can also set it manually within Xcode even with String Catalogs have more languages in them – I might not be ready to use them all, some maybe be WIP. So, how can I set this setting here:

<img width="801" alt="Bildschirmfoto 2024-08-07 um 11 14 56" src="https://github.com/user-attachments/assets/fc77451e-656e-462e-b3ed-f23816706dba">

**Comment by Jeehut** [comment]
UPDATE: We just found that there's the [`defaultKnownRegions`](https://docs.tuist.io/references/project-description/structs/project.options#defaultknownregions) option we can set to get the languages we want. I have no idea why it has the `default` prefix, which seems very confusing, but it fulfills my needs.

With this, I personally consider Tuist compatible with String Catalogs and I don't think any further steps are necessary. Except if you want to provide generated references for String Keys marked as "Migrated" inside a String Catalog, which Xcode marks those keys that were migrated from Strings(dict) files. Or maybe the docs should be updated to clarify things.

**Comment by leszko11** [comment]
From my point of view, the most important thing about this implementation is having a generated structure of keys existing in the Strings Catalog. Calling `L10n.Your.Key.name` is more convenient, and errorproof than providing the key manually. If I do a typo, the fact, that String Catalog will have this incorrect key, does not change a thing. And this is what you wrote in a linked blogpost
> Note that you won't get auto-completion for your keys in code like you now get for images and colors in your Asset Catalogs with Xcode 15

And looking at replies/reactions [here](https://github.com/SwiftGen/SwiftGen/issues/1065) and [there](https://github.com/SwiftGen/SwiftGen/pull/1124) developers like the auto-completion

**Comment by Jeehut** [comment]
I don't think you got my point. Calling `L10n.Your.Key.name` for sure is NOT more convenient than just writing "Hello World" into the SwiftUI view. Nothing is more convenient than that. And Yes, with String Catalogs, writing "Hello World" into your SwiftUI view IS in fact error-proof. There's no way Xcode will miss to add that to the String Catalog as long as you're using localized APIs (like `LocalizedStringKey` or `String(localized:)`). In other words, there's no longer need for auto-completion anymore. Auto-completion of what anyway? The keys in the String Catalogs are the _output_ of writing code, no longer the source. The whole "source of truth" has been reversed. People don't understand this, therefore they ask for auto-completion. I was like that too in the first moment. But now I understand String Catalogs better.

And your quote from my blog post is deliberately incomplete, because right in the next sentence I explain why. If you quote me, please do it correctly. Here's the full quote:

> Note that you won't get auto-completion for your keys in code like you now get for images and colors in your Asset Catalogs with Xcode 15. That's because in Asset Catalogs the source of truth lies in the catalog itself, where your assets are placed. But because Xcode automatically extracts any added localizations from your source code, the source of truth for your localizations is reversed here and lies in your code. Therefore it's impossible that you add a text to your code that lacks a counterpart in a Strings file (like it was possible before, leading to broken translations). Instead, whenever you add a localized String in your project, Xcode automatically creates a new translation key for all supported languages and you'll see in the Strings Catalog that your translations are incomplete.

**Comment by h8wait** [comment]
Hey @Jeehut, let me join the discussion. First of all, thanks for your messages—they explain the paradigm of working with string catalogs very well.

That being said, I can't agree with your statement that Tuist is already compatible with string catalogs, and here's why.

It seems to me that you are looking at the problem too narrowly. Imagine that your project is powered by Tuist and is already using the `L10n.Your.Key.name` semantic and `.strings` legacy localization. Following your logic, it seems that we would just need to migrate the `.strings` files, apply `defaultKnownRegions`, and call it a day. The problem only begins there, because we would need to replace all the old code-generated semantics like `L10n.Your.Key.name` with the native `String(localized:)`. Even if some automation can handle this, it could result in a huge pile of changes for large projects.

For the private project that I'm working on right now, it would involve a pull request with approximately 3k changes that would require approvals from code owners—basically from every team member. In my opinion, this is not how the migration should look. I would prefer to have the ability to change the localization source without making massive changes to the domain-level codebase. After that, the rule to use the native approach can be established, allowing each team to migrate the paradigm iteratively within its own domain.

**Comment by Jeehut** [comment]
I do understand that old code can't be migrated if you are using SwiftGen-style generated code in your code rather than Apples native APIs. But, if you're using native APIs like NSLocalizedString then you can totally migrate and use Tuist with String Catalogs. And if you start a new project, you can also use String Catalogs.

So, what you are asking for is for Tuist to provide a migratory from SwiftGen-based code references back to native APIs. That's not the same as generally supporting String Catalogs.

And even in that situation, you could totally keep your existing translations in your existing String files and just start adding new keys with the native APIs. I have not tested if Xcode can handle a mix of String Catalogs and Strings Files with the same name, but if not, it should be easy enough to just rename the Strings file to something like LegacyLocalizable.strings, adjust the SwiftGen config to specify this tableName in the generated code and migrate to String Catalogs only for new code.

**Comment by Jeehut** [comment]
> The problem only begins there, because we would need to replace all the old code-generated semantics like L10n.Your.Key.name with the native String(localized:).

Why would you need to replace all the old code-generated semantic keys? Just migrate your Strings file to a String Catalog, remove SwiftGen as a build step, but KEEP the generated file in its latest form. As long as you don't remove any migrated keys from the String Catalog, all should build just fine. And from there on, you would use native APIs for new code you add. And optionally, over time, if you really want to, then yes, could also migrate the L10n calls to native APIs.

But I would hold off on that until somebody has provided a migrator. But that's a separate task that could be fixed in its own ticket here on Tuist, or even better, in SwiftGen. That's how I would approach it. But I still think Tuist supports Strinf Catalogs already.

**Comment by h8wait** [comment]
> Just migrate your Strings file to a String Catalog, remove SwiftGen as a build step, but KEEP the generated file in its latest form.

Wdym by keep? The entire Tuist `Derived` folder is gitignored, including the file with semantic keys. This means that we would need to ADD this giant file to source control and deal with potential merge conflicts since we would no longer be using code generation.

How long do you think this file will remain in the codebase? I think for a real-world project, it might survive for a year or so, depending on the team's priorities. Once again, this is not how I believe the migration should look.

**Comment by Jeehut** [comment]
> Wdym by keep? The entire Tuist Derived folder is gitignored, including the file with semantic keys. This means that we would need to ADD this giant file to source control and deal with potential merge conflicts since we would no longer be using code generation.

Oh, right, the file isn't checked in, I forgot that. But then, yeah, just add it to your project and commit. You don't need to worry about merge conflicts because that file shouldn't be edited by anyone from here on, it's just a legacy compatibility file.

All future changes should happen within the String Catalog. Which has a JSON structure and sorts its keys alphabetically, so the merge conflicts for that file should be minimal.

**Comment by leszko11** [comment]
I still see a lot of value in generating keys list regardless of the source. And I believe I'm not the only one. This feature was requested here, in Swiftgen, and there's also quite popular [build plugin](https://github.com/liamnichols/xcstrings-tool) doing the same thing.

But it's your call @pepicrft

**Comment by bogren** [comment]
I really like the change Apple is making here, changing the way we work with localization, and appreciate the recent input on this topic! The paradigm shift is very welcomed, and I think most should use this for their projects. 

Unfortunately, this doesn't work for everyone and every project. Instead there is, like @leszko11 said, great value in generating the strings and continue working with strings backwards from the generated file. We work on rather large project where we don't have the luxury to allow the code to be the truth. I can see any other project with external agency for translation, multi-platform, any legacy reasons, or other restrictions have the need for generating keys regardless of the source.

**Comment by Iron-Ham** [comment]
There is incredible value in generating the strings keys. Let's take a look at a basic example string separating a few names out. 

Input: [UserA]
Output: "UserA"

Input: [UserA, UserB]
Output: "UserA, UserB"

Input: [UserA, UserB, UserC]
Output: "UserA, UserB, and 1 other."

Input: [UserA, UserB, UserC, UserD]
Output: "UserA, UserB, and 2 others."

This is a common UI paradigm. Historically, this was split into two strings: The base case ("UserA"), and the pluralistic case. For this discussion, let's only look at the plural case. 

In the new `xcstrings` files, we'd define something like this `%@ %@ %d others`, with a key value of `%1$@ %2$@ %lld others". That would then get translated in to the various cases. Here's the thing, whether you're using SwiftUI or UIKit, this is not in any way shape or form understandable:

```swift
Text("\(userA) \(userB) \(otherUserCount) others") // Wtf should this show? Is this intended?
```

Compare that to:

```swift
Text(L10n.dOthers(userA, userB, otherUserCount)) // makes it **super obvious** that the value of this is derived from an asset catalog
```

The conversation around simple strings like `Hello World` miss the point entirely. Some apps have thousands – or _tens of thousands_  – of complex rule-matched strings with various replacement keys and so on. These strings cannot be clearly represented as a blob of text, and not knowing whether something is a naked string or an intentionally formatted localized value is a huge issue in these codebases.

**Comment by Jeehut** [comment]
@Iron-Ham I personally do not think your `L10n` example is in any way more readable than the plain `Text` example above it. How is it more obvious? It's literally exactly the same amount of information. If anything, you would need something like `L10n.listOfUsersMoreThanTwo(...)` which is more readable.

But if your concern is that you want to have "semantic keys", that's something I'm planning to solve with a simple `#L10n` macro at some point. I just didn't have the time to do that yet, but it's gonna compile to a plain `String(localized:)` while allowing to use semantic keys. There is already an API for that, but it's quite cumbersome, so a perfect use for a macro. Just didn't have the time to write it yet, busy with other stuff.

**Comment by pepicrft** [comment]
👋🏼 
Trying to distil the points int he above comments:
1. Apple is proposing a new paradigm.
2. But some projects can't afford the paradigm shift.

I don't quite understand why not everyone can afford the shift. Based on the discussions, it seems there are some limitations in the new APIs and potentially some subjective preferences. Which leads me to the question, which role should Tuist play here?

1. Look ahead and support Apple's new model filling the gaps like @Jeehut described (e.g. bringing some ergonomy in some areas)
2. Continue to support the old model
3. Support companies in the transition from the old paradigm to the new one.

**Comment by CaramelHeaven** [comment]
In my humble opinion, why should we decide what is right and what is wrong, can we provide a tool that will make both sides happy? Add support for the old model and the new model, and let the user decide what is best for him in the current situation of the project

**Comment by leszko11** [comment]
I believe that Strings catalog existence, and the way how it works, doesn't eliminate the need for resource generation in Tuist, which is still essential for improving developers workflows and ensuring consistency across projects, especially when handling large or complex localization setups.

It's also worth noticing that the majority of contributors in this thread seem to agree that this feature would add value to the Tuist. I appreciate that @Jeehut has raised valid concerns about this feature is necessary, but I feel that the concensus points towards its usefulness for many users.

**Comment by daniil-konashenko** [comment]
Another reason why I support @leszko11's opinion is that I am working on a medium-sized project. I have noticed that the 'Syncing Localizations' phase, which is supposed to occur during every build, does not always function correctly and the directory does not get updated. Because of this, I simply cannot trust the code to be accurate, as Apple suggests. I would greatly appreciate it if the feature that @leszko11 has been working on could be included in the next release.

**Comment by pepicrft** [comment]
Alright. @leszko11 what would you need from us to push this work to the finish line?

**Comment by leszko11** [comment]
We need this [PR](https://github.com/SwiftGen/SwiftGen/pull/1124) in SwiftGen repo to be merged

**Comment by rakuyoMo** [comment]
@pepicrft Can Tuist allow users to modify the path of the built-in SwiftGen? For example, provide an environment variable named `TUIST_SWIFT_GEN_PATH` (or other names). By setting this variable, users can let Tuist use the user-compiled version of SwiftGen instead of the built-in version.

**Comment by fortmarek** [comment]
> Can Tuist allow users to modify the path of the built-in SwiftGen? For example, provide an environment variable named TUIST_SWIFT_GEN_PATH (or other names).

No, primarily because we don't use SwiftGen as a CLI but as a package dependency: https://github.com/tuist/tuist/blob/main/Package.swift#L502

However, at this point, I think we'd accept a PR that uses a `SwiftGen` fork given the inactivity from the `SwiftGen` maintainers.

**Comment by rakuyoMo** [comment]
> No, primarily because we don't use SwiftGen as a CLI but as a package dependency: https://github.com/tuist/tuist/blob/main/Package.swift#L502
> 
> However, at this point, I think we'd accept a PR that uses a `SwiftGen` fork given the inactivity from the `SwiftGen` maintainers.

Thanks for your reply, I'm glad to see new progress on this issue.

But whose fork to use is a question. Will Tuist officially fork? This may cause the SwiftGen community to lean towards Tuist/SwiftGen, which may increase Tuist's maintenance costs.

**Comment by fortmarek** [comment]
> Will Tuist officially fork? This may cause the SwiftGen community to lean towards Tuist/SwiftGen, which may increase Tuist's maintenance costs.

There are three options to consider:
- We keep waiting to see if SwiftGen repo becomes alive.
- We fork `SwiftGen` under `Tuist` with possibly increasing our maintenance burden.
- We use @leszko11's fork.

@leszko11 how do you feel about maintaining a fork until (if ever) `SwiftGen` becomes active again? We can also fork it under the Tuist org and give you the permission to maintain it – or we can use your personal organization.

**Comment by leszko11** [comment]
Hey!

Seeing other people's comments/reactions in all related issues and PRs makes me wanna (at least try) to bring back `SwiftGen` alive.

@fortmarek I'd choose an option where I'll maintain a fork under `Tuist` org. However, I'll follow your suggestion on the most suitable option to proceed.

**Comment by fortmarek** [comment]
Hey @leszko11 👋 

I created a fork at https://github.com/tuist/SwiftGen and made you a maintainer of that repository. I'm happy to help with reviews and anything else you might need 🙏 Thanks for taking this on.

**Comment by leszko11** [comment]
@fortmarek I should wrap-up a PR by the end of this week.
I would appreciate any kind of help regarding reviews, releasing, and everything else 🙌

**Comment by pepicrft** [comment]
Thanks a lot @leszko11. Can't wait for that PR.

**Comment by CaramelHeaven** [comment]
@fortmarek Hi!
Would it be possible to review and approve this MR?
https://github.com/tuist/SwiftGen/pull/1

It adds support for .xcstrings, and we — as developers — are really looking forward to this feature to improve how we handle localization strings via the catalog.
Thanks a lot in advance!

**Comment by pepicrft** [comment]
Hey folks!
Apologies for it. I'll try to prioritize it before the end of the week.

**Comment by freak4pc** [comment]
It seems SwiftGen#1 was merged but still no support in Tuist for xcstrings?

**Comment by leszko11** [comment]
Hey @freak4pc 

I'll try to wrap it up this week

**Comment by mkemalgokce** [comment]
is there any update for this?

**Comment by CaramelHeaven** [comment]
Any update?

**Comment by vldalx** [comment]
Maybe my comment [here](https://github.com/SwiftGen/SwiftGen/pull/1124#issuecomment-3927946716) will be useful to someone

**Comment by Athx9** [comment]
I have experience with Tuist resource mappers and would like to implement support for .xcstrings catalogs. Please assign this to me.

**Comment by devman247** [comment]
Hi, I am interested in solving this issue. Please send me an invitation to join.
